pkg/controller/operators/adoption_controller.go

```diff
@@ -0,0 +1,409 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	operatorsv2alpha1 "github.com/operator-framework/api/pkg/operators/v2alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/operators/decorators"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
+)
+
+// AdoptionReconciler automagically associates Operator components with their respective operator resource.
+type AdoptionReconciler struct {
+	client.Client
+
+	log     logr.Logger
+	mu      sync.RWMutex
+	factory decorators.OperatorFactory
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=operators,verbs=create;update;patch;delete
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=operators/status,verbs=update;patch
+// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch
+
+// SetupWithManager adds the operator reconciler to the given controller manager.
+func (r *AdoptionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Trigger operator events from the events of their compoenents.
+	enqueueSub := &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.mapToSubscriptions),
+	}
+
+	// Create multiple controllers for resource types that require automatic adoption
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&operatorsv1alpha1.Subscription{}).
+		Watches(&source.Kind{Type: &operatorsv1alpha1.ClusterServiceVersion{}}, enqueueSub).
+		Complete(reconcile.Func(r.ReconcileSubscription))
+	if err != nil {
+		return err
+	}
+
+	enqueueCSV := &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.mapToClusterServiceVersions),
+	}
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&operatorsv1alpha1.ClusterServiceVersion{}).
+		Watches(&source.Kind{Type: &appsv1.Deployment{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Namespace{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Service{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.ServiceAccount{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.ConfigMap{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.Role{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.RoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRole{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &apiextensionsv1.CustomResourceDefinition{}}, enqueueCSV).
```

Comment:
should we put watch on `apiextensionsv1beta1.CustomResourceDefinition` types as well?

Reply:
you'd end up with duplicate resources since a list returns _all_ CRDs converted to the request version.